File: 1-10/3_Largest_prime_factor/main.c
/*
the answer 6857
*/
//2014.09.07//
//2014.11.06 pdfŒ`®‚É•ÏX//
#include<stdio.h>
#include<math.h>
#define NUM 600851475143
int main(){
	double num = NUM;
	int factor = 2,
		 lastFactor,
		 maxFactor,
		 i;

	//2‚ÅŠ„‚é
	if(fmodl(num,factor) == 0)
	{
			lastFactor = factor;
			num /= factor;
			while(fmodl(num,factor) == 0)
				num /= factor;
	}else
		lastFactor = 1;

	factor = 3;
	maxFactor = sqrt(num);
	while(num > 1 && factor <= maxFactor)
	{
		if(fmodl(num,factor) == 0)
		{
			num /= factor;
			lastFactor = factor;
			while(fmodl(num,factor) == 0)
				num /= factor;
			maxFactor = sqrt(num);
		}
		factor += 2;
	}
	if(num == 1)
		printf("%d\n",lastFactor);
	else
		printf("%.0f", num);
	getchar();
	return 0;
}

File: 1-10/3_Largest_prime_factor/main.py
# coding:utf-8
# 13195 の素因数は 5, 7, 13, 29 である.
# 600851475143 の素因数のうち最大のものを求めよ.

from math import sqrt

def devideN(number, factor):
	number //= factor
	while number % factor == 0:
		number //= factor;
	return number

num = N = 600851475143
maxFactor = 0

if num % 2 == 0:
	num = devideN(num, 2)

factor = 3
maxFactor = sqrt(num)

while num > 1 and factor <= maxFactor:
	if num % factor == 0:
		num = devideN(num, factor)
		maxFactor = sqrt(num)
	factor += 2

print(N, "の素因数の中で最大のものは", num, "です")